windows/src/engine/keyman32/appint/appint.cpp

```diff
@@ -107,16 +107,40 @@ void AppContext::Reset()
 
 void AppContext::Get(WCHAR *buf, int bufsize)
 {
-	for(WCHAR *p = CurContext; *p && bufsize > 0; p++, bufsize--)
-	{
-		*buf = *p; buf++;
-		if(*p >= 0xD800 && *p <= 0xDBFF) { *buf = *(++p); bufsize--; buf++; }
-	}
-	//wcsncpy(buf, CurContext, bufsize);
-	*buf = 0;
-	//buf[bufsize-1] = 0;
+  // surrogate pairs need to be treated as a single unit, therefore use
+  // BuffMax to find the startIndex.
+  // Buff Max handles the case where the surrogate pairs is split
+  // by the buff size
+  WCHAR* startIndex = this->BufMax(bufsize);
+
+  for (WCHAR* p = startIndex; *p && bufsize > 0; p++, bufsize--)
+  {
+    *buf = *p;
+    if ((*p >= 0xD800 && *p <= 0xDBFF) && (bufsize - 2 > 0)) { buf++;  *buf = *(++p); bufsize--; }
```

Comment:
This is making the assumption that we don't have an unpaired surrogate in the buffer. Is that a harmful assumption? I don't think so.

Reply:
```suggestion
    if(Uni_IsSurrogate1(*p) && bufsize - 2 > 0) { 
      buf++; p++;
      *buf = *p;
      bufsize--;
    }
```

Reply:
> This is making the assumption that we don't have an unpaired surrogate in the buffer. Is that a harmful assumption? I don't think so.
Nice edit.  I don't think it is a harmful assumption as it will still copy an unpaired surrogate (it will not skip or anything). If Get returned an error code it would be a different story.